src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6423,6 +6448,79 @@ void reinferMethodAndVisitArguments(BoundCall node, TypeWithState receiverType,
                 SetResult(node, returnState, method.ReturnTypeWithAnnotations);
                 SetUpdatedSymbol(node, node.Method, method);
             }
+
+            static ImmutableArray<BoundExpression> getArguments(ImmutableArray<BoundExpression> arguments, bool isNewExtension, BoundExpression? receiver)
+            {
+                // In error cases, the receiver may have already been stored as an argument
+                if (isNewExtension && receiver is not null)
+                {
+                    return [receiver, .. arguments];
+                }
+
+                return arguments;
+            }
+
+            static ImmutableArray<ParameterSymbol> getParameters(ImmutableArray<ParameterSymbol> parameters, bool isNewExtension, MethodSymbol method)
+            {
+                if (!isNewExtension)
+                {
+                    return parameters;
+                }
+
+                ParameterSymbol? extensionParameter = method.ContainingType.ExtensionParameter;
+                Debug.Assert(extensionParameter is not null);
+
+                return [extensionParameter, .. parameters];
+            }
+        }
+
+        private static ImmutableArray<RefKind> GetArgumentRefKinds(ImmutableArray<RefKind> argumentRefKindsOpt, bool isNewExtension,
+            MethodSymbol method, int argumentCount)
+        {
+            if (!isNewExtension)
+            {
+                return argumentRefKindsOpt;
+            }
+
+            ParameterSymbol? extensionParameter = method.ContainingType.ExtensionParameter;
+            Debug.Assert(extensionParameter is not null);
+            var receiverRefKind = extensionParameter.RefKind == RefKind.Ref ? RefKind.Ref : RefKind.None;
+
+            if (argumentRefKindsOpt.IsDefault)
+            {
+                if (extensionParameter.RefKind == RefKind.None)
```

Comment:
Shouldn't this check `receiverRefKind`? Otherwise if `extensionParameter.RefKind` is `In` for example, we construct the array unnecessarily.